core/app/models/spree/address.rb

```diff
@@ -218,6 +228,12 @@ def validate_name
     end
 
     def state_validate
+      Spree::Deprecation.warn \
```

Comment:
Until the preference is not changed users will see 3 deprecation warnings. It would be awesome to find a way to only print one deprecation warning in that case, but keep the individual ones if these methods are called from somewhere else. I'm not sure it's a good idea but what about passing an arg to these deprecated methods to conditionally silence the deprecation warning?

This way we can do something like:

```
      if Spree::Config.legacy_address_state_validator
        state_validate(silence_deprecation: true)
        validate_state_matches_country(silence_deprecation: true)
      else
```

Users still see the main deprecation warning if using default code and the specific ones if they used `state_validate` or `validate_state_matches_country` in other ways. I'm not particularly concerned about changing the arity of those methods since they'll go away soon. 

What's your thoughts?

Reply:
👍 agree!
By introducing a new parameter it will break any existing prepended patches of these methods. I'm not aware of a way to introduce a new parameter w/o being breaking. I'm looking if there's another way

Reply:
Maybe just setting an accessor on the class?

Reply:
Another way might be to wrap the two calls in a `silence` block, this avoids polluting the Address class API with new attributes/methods. The downside is it will silence any eventual deprecations generated by these two calls.
What do you think?

```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    Spree::Deprecation.silence do
      state_validate
      validate_state_matches_country
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```

Reply:
It's more elegant for sure but I think (might be worth a double-check) we are not using `Spree::Deprecation.silence` anywhere in the codebase (outside specs). I'm not completely sure why to be honest and if this could cause issues that we can't anticipate at the moment. Anyway, for now this seems like the best solution let's go head with this and let's see what the other core team member think about it.

Reply:
just discovered that `silence` seems to not be thread-safe 🤦 Here's a simple test case that sometimes leaves silenced class attribute set to `true`: https://gist.github.com/cedum/8315ebdd59781ee7fcc7f452a405542f
I'm refactoring this to a simple instance variable switch 😄, something like this:
```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    begin
      @silence_state_deprecations = true
      state_validate
      validate_state_matches_country
    ensure
      @silence_state_deprecations = false
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```